### PR TITLE
Added memory recycling and optimized writes

### PR DESCRIPTION
**Pull request by AlasdairF**
I added memory pooling so buffers do not have to be reallocated.
I also optimized the Write & Encode functions so that they will now usually perform only 1 write to the underlying writer, instead of 2.
These optimizations will drastically speed up the reading and writing at the cost of around 150KB memory overhead to the user's application.
I also added some more constants to avoid unnecessary calculations.

### CODE REVIEW

File: snappy.go [hunk 1]
```diff
@@ -6,7 +6,7 @@
 // It aims for very high speeds and reasonable compression.
 //
 // The C++ snappy implementation is at https://github.com/google/snappy
-package snappy // import "github.com/golang/snappy"
```
Comment (cespare) [inline snappy.go]: Why did you remove the import comment?

Comment (AlasdairF) [inline snappy.go]: I removed the import comment only because Go would not get the repository unless I did this.

**Comment by cespare** [comment]
I doubt a change adding an external dependency will be accepted, especially when the dependency is only a tiny amount of code and doesn't have a license. (Also, sync.Pool might make more sense.)

File: snappy.go (hunk 1 above)
Comment (cespare) [inline snappy.go]: If you're working on a local fork, you can check that out manually to the correct location (`$GOPATH/src/github.com/golang/snappy`) even if the remote is e.g. github.com/AlasdairF/snappy. `go get` is a convenience; sometimes it doesn't fit your use case and you need to use `git clone`.

This import comment is intentional, and should be left in.

**Comment by AlasdairF** [comment]
Sure, these changes are for my own use really - I don't actually expect it to actually be approved. Still... it's good for everyone to see these optimizations. Reallocating buffers and unnecessary writes are never a good idea.

File: snappy.go (hunk 1 above)
Comment (dgryski) [inline snappy.go]: A good tutorial for this is https://splice.com/blog/contributing-open-source-git-repositories-go/

**Comment by AlasdairF** [comment]
I'll do as suggested and create a new pull request.
